internal/cmd/protoc-gen-go-flipt-sdk/main.go

```diff
@@ -303,10 +313,96 @@ func (p StaticTokenAuthenticationProvider) Authentication() (string, error) {
 type JWTAuthenticationProvider string
 
 // Authentication returns the underlying string that is the JWTAuthenticationProvider.
-func (p JWTAuthenticationProvider) Authentication() (string, error) {
+func (p JWTAuthenticationProvider) Authentication(context.Context) (string, error) {
     return "JWT " + string(p), nil
 }
 
+// KubernetesAuthenticationProvider is an implementation of ClientAuthenticationProvider
+// which automatically uses the service account token from the environment and exchanges
+// it with Flipt for a client token.
+// This provider keeps the client token up to date and refreshes it for a new client
+// token before expiry. It re-reads the service account token as Kubernetes can and will refresh
+// this token, as it also has its own expiry.
+type KubernetesAuthenticationProvider struct {
+	transport               Transport
+	serviceAccountTokenPath string
+	leeway                  time.Duration
+
+	mu   sync.RWMutex
+	resp *auth.VerifyServiceAccountResponse
+}
+
+// KubernetesAuthenticationProviderOption is a functional option for configuring KubernetesAuthenticationProvider.
+type KubernetesAuthenticationProviderOption func(*KubernetesAuthenticationProvider)
+
+// WithKubernetesServiceAccountTokenPath sets the path on the host to locate the kubernetes service account.
+// The KubernetesAuthenticationProvider uses the default location set by Kubernetes.
+// This option lets you override that if your path happens to differ.
+func WithKubernetesServiceAccountTokenPath(p string) KubernetesAuthenticationProviderOption {
+	return func(kctp *KubernetesAuthenticationProvider) {
+		kctp.serviceAccountTokenPath = p
+	}
+}
+
+// WithKubernetesExpiryLeeway configures the duration leeway for deciding when to refresh
+// the client token. The default is 10 seconds, which ensures that tokens are automatically refreshed
+// when their is less that 10 seconds of lifetime left on the previously fetched client token.
+func WithKubernetesExpiryLeeway(d time.Duration) KubernetesAuthenticationProviderOption {
+	return func(kctp *KubernetesAuthenticationProvider) {
+		kctp.leeway = d
+	}
+}
```

Comment:
I wonder if we should have restrictions for the value of this `leeway`?

Reply:
I think we might be ok with just a sensible default like it has here. I hope no one really has to mess with this.
They can't really do much damage by setting it innapropriately either. At worst they will request a new token on every request (if leeway > duration of lifetime), or they may encounter the odd expired error if they set it too low.
I think any bounds we pick here might be arbitrary, im not sure.